src/approaches/grammar_search_invention_approach.py

```diff
@@ -373,6 +381,42 @@ def enumerate(self) -> Iterator[Tuple[Predicate, float]]:
             yield (predicate, cost)
 
 
+@dataclass(frozen=True, eq=False, repr=False)
+class _PrunedGrammar(_DataBasedPredicateGrammar):
+    """A grammar that prunes redundant predicates.
+    """
+    base_grammar: _PredicateGrammar
+
+    def enumerate(self) -> Iterator[Tuple[Predicate, float]]:
+        # Predicates are identified based on their evaluation across
+        # all states in the dataset.
+        seen = {}  # maps identifier to previous predicate, for debugging
```

Comment:
Why "for debugging"? Isn't this necessary for checking redundancy, which is the whole point of this class?